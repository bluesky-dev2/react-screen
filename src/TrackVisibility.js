/* global window, document */
import React, { Component, PropTypes } from 'react';
import throttle from 'lodash.throttle';

const propTypes = {
  once: PropTypes.bool,
  throttleInterval(props, propName, component) {
    const currentProp = props[propName];
    if (!Number.isInteger(currentProp) || currentProp < 0) {
      return new Error(`The ${propName} prop you provided to ${component} is not a valid integer >= 0.`);
    }
  },
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element)
  ]),
  style: PropTypes.object,
  className: PropTypes.string
};

const defaultProps = {
  once: false,
  throttleInterval: 150,
  style: null,
  className: null
};

export default class TrackVisibility extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isVisible: false
    };
    this.isComponentVisible = this.isComponentVisible.bind(this);
    /* Store reference to be able to remove the event listener */
    this.throttleCb = throttle(this.isComponentVisible, this.props.throttleInterval);
    this.setNodeRef = ref => { this.nodeRef = ref; };
  }

  componentDidMount() {
    this.attachListener();
    this.isComponentVisible();
  }

  componentWillUnmount() {
    this.removeListener();
  }

  getPropsToRender() {
    const props = {};

    this.props.className !== null && (props.className = this.props.className);
    this.props.style !== null && (props.style = this.props.style);

    return props;
  }

  getChildProps() {
    const props = {};
    Object.keys(this.props).forEach(key => {
      if (!{}.hasOwnProperty.call(TrackVisibility.defaultProps, key)) {
        props[key] = this.props[key];
      }
    });
    return props;
  }

  getChildren() {
    return React.Children.map(
      this.props.children,
      child => React.cloneElement(child, { ...this.getChildProps(), isVisible: this.state.isVisible })
    );
  }

  isComponentVisible() {
    const rect = this.nodeRef.getBoundingClientRect();
    const html = document.documentElement;

    if (
      rect.top >= 0 &&
      rect.left >= 0 &&
      rect.bottom <= (window.innerHeight || html.clientHeight) &&
      rect.right <= (window.innerWidth || html.clientWidth)
    ) {
      this.props.once && this.removeListener();
      !this.state.isVisible && this.setState({ isVisible: true });
    } else {
      this.state.isVisible && this.setState({ isVisible: false });
    }
  }

  attachListener() {
    window.addEventListener('scroll', this.throttleCb);
    window.addEventListener('resize', this.throttleCb);
  }

  removeListener() {
    window.removeEventListener('scroll', this.throttleCb);
    window.removeEventListener('resize', this.throttleCb);
  }

  render() {
    return (
      <div ref={this.setNodeRef} {...this.getPropsToRender()}>
        {this.getChildren()}
      </div>
    );
  }
}

TrackVisibility.propTypes = propTypes;
TrackVisibility.defaultProps = defaultProps;
