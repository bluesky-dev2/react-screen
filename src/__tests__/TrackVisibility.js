import React from 'react';
import sinon from 'sinon';
import TrackVisibility from './../TrackVisibility';


const hasProp = (Component, prop) => {
 return {}.hasOwnProperty.call(Component, prop);
};

const Dumb = props => <div>Plop</div>;

describe('<TrackVisibility />', () => {

  describe('When rendering the component', ()=> {
    it('Can be rendered without children', () => {
      const wrapper = mount(<TrackVisibility />);
      expect(wrapper).toMatchSnapshot();
    });

    it('Should bind event to window on mount', () => {
      window.addEventListener = jest.fn();    
      mount(<TrackVisibility />);
      expect(window.addEventListener).toHaveBeenCalled();
    });

    it('Should remove event from window on unmount', () => {
      window.removeEventListener = jest.fn();    
      const wrapper = mount(<TrackVisibility />);
      wrapper.unmount();
      expect(window.removeEventListener).toHaveBeenCalled();
    });

    it('Allows us to set props', () => {
      const wrapper = mount(<TrackVisibility bar="baz" />);
      expect(wrapper.props().bar).toBe("baz");
      wrapper.setProps({ bar: "foo" });
      expect(wrapper.props().bar).toBe("foo");
    });

    it('Throws an error if throttleInterval is not a valid integer >= 0', () => {
      const stub = sinon.stub(console, 'error');
      [-1, 'dsqdqsqs', 1.10].forEach((v) => {
        mount(<TrackVisibility throttleInterval={v}/>);
        expect(stub.calledOnce).toBe(true);
      });
      console.error.restore();
    });

    it('Should pass isVisible to children as prop', () => {
      const wrapper = mount(
        <TrackVisibility>
          <Dumb />
        </TrackVisibility>
      );
      expect(hasProp(wrapper.find('Dumb').props(), 'isVisible')).toBe(true);
    });

    it('Should transfer all the props provided to children but omit own props [once, throttleInterval, className, style]', () => {
      const wrapper = mount(
        <TrackVisibility foo="bar" baz="foobar">
          <Dumb />
        </TrackVisibility>
      );
      const props = wrapper.find('Dumb').props();
      
      expect(hasProp(props, 'foo')).toBe(true);
      expect(hasProp(props, 'baz')).toBe(true);
      expect(hasProp(props, 'isVisible')).toBe(true);

      expect(hasProp(props, 'className')).toBe(false);
      expect(hasProp(props, 'style')).toBe(false);
      expect(hasProp(props, 'once')).toBe(false);
      expect(hasProp(props, 'throttleInterval')).toBe(false);
    });

    it('Can update state', () => {
      const wrapper = mount(<TrackVisibility />);
      wrapper.setState({isVisible: true});
      expect(wrapper.state('isVisible')).toBe(true);
      wrapper.setState({isVisible: false});
      expect(wrapper.state('isVisible')).toBe(false);
    });

    // it('Should have offset prop if set', () => {
    //   const wrapper = shallow(<TrackVisibility offset={500} />);
    //   const props = wrapper.props();
    //   expect(hasProp(props, 'offset')).to.equal(true);
    //   expect(props.offset).to.equal(500);
    // });
  });
});
