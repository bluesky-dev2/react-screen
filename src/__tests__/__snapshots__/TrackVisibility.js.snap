// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackVisibility /> When rendering the component Can be rendered without children 1`] = `
<TrackVisibility
  offset={0}
  once={false}
  partialVisibility={false}
  tag="div"
  throttleInterval={150}
>
  <div />
</TrackVisibility>
`;

exports[`<TrackVisibility /> When rendering the component Should be able to use a render props 1`] = `
<TrackVisibility
  offset={0}
  once={false}
  partialVisibility={false}
  tag="div"
  throttleInterval={150}
>
  <div>
    <Dumb>
      <div>
        Plop
      </div>
    </Dumb>
  </div>
</TrackVisibility>
`;

exports[`<TrackVisibility /> When rendering the component Should fallback clientHeight when window propeties are not defined 1`] = `
<TrackVisibility
  offset={0}
  once={false}
  partialVisibility={false}
  tag="div"
  throttleInterval={150}
>
  <div />
</TrackVisibility>
`;

exports[`<TrackVisibility /> when partially visible Can check partial visibilty 1`] = `
<TrackVisibility
  offset={0}
  once={false}
  partialVisibility={true}
  tag="div"
  throttleInterval={150}
>
  <div />
</TrackVisibility>
`;
